internal/worker/firewaller/firewaller.go

```diff
@@ -1501,171 +1518,279 @@ func (ad *applicationData) scopedContext() (context.Context, context.CancelFunc)
-		// placeholder
-		_ = fw.startRelation(ctx, &params.RemoteRelation{
-			RemoteApplicationName: "",
-		}, "")
+		return true, rel, nil
 	}
-	return nil
+	return false, rel, nil
 }
 
-type remoteRelationInfo struct {
-	relationToken string
-}
+// startConsumerRelation creates a watcher for egress changes on the consuming
+// side. The behavior depends on the local endpoint role:
+// - If requirer: watch local egress and publish to remote.
+// - If provider: watch remote egress and apply ingress locally.
+func (fw *Firewaller) startConsumerRelation(ctx context.Context, rel domainrelation.RelationDetails) error {
+	tag := names.NewRelationTag(rel.Key.String())
+	fw.logger.Debugf(ctx, "starting consumer relation watcher for %v", tag.Id())
 
-type remoteRelationData struct {
-	catacomb      catacomb.Catacomb
-	fw            *Firewaller
-	relationReady chan remoteRelationInfo
+	// For consuming side, we need to identify which endpoint is local and which
+	// is remote.
+	if len(rel.Endpoints) != 2 {
+		return errors.Errorf("relation %v has %d endpoints, expected 2", tag.Id(), len(rel.Endpoints))
+	}
 
-	tag                 names.RelationTag
-	localApplicationTag names.ApplicationTag
-	relationToken       string
-	remoteModelUUID     string
-	endpointRole        charm.RelationRole
-	isOffer             bool
+	// Try to identify the remote application by querying both endpoints.
+	var localEndpoint domainrelation.Endpoint
+	var remoteApplicationName string
+	var remoteModelUUID string
 
-	crossModelFirewallerFacade CrossModelFirewallerFacadeCloser
+	for _, ep := range rel.Endpoints {
+		remoteApps, err := fw.crossModelRelationService.RemoteApplications(ctx, []string{ep.ApplicationName})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(remoteApps) > 0 && remoteApps[0].Result != nil {
+			// This is the remote application.
+			remoteApplicationName = ep.ApplicationName
+			remoteModelUUID = remoteApps[0].Result.ModelUUID
+		} else {
+			// This is the local application.
+			localEndpoint = ep
+		}
+	}
 
-	// These values are updated when ingress information on the
-	// relation changes in the model.
-	ingressRequired bool
-	networks        set.Strings
+	if remoteApplicationName == "" {
+		return errors.Errorf("could not identify remote application for relation %v", tag.Id())
+	}
+
+	data := &remoteRelationData{
+		fw:                  fw,
+		tag:                 tag,
+		localApplicationTag: names.NewApplicationTag(localEndpoint.ApplicationName),
+		relationUUID:        rel.UUID,
+		remoteModelUUID:     remoteModelUUID,
+	}
+	var watchEgressLoop func() error
+	if localEndpoint.Role == charm.RoleRequirer {
+		// Local endpoint is requirer, so watch local egress and publish to
+		// remote (offering model).
+		watchEgressLoop = data.watchLocalEgressPublishRemote
+	} else {
+		// Local endpoint is provider, so watch remote (offering model) egress
+		// and apply locally.
+		watchEgressLoop = data.watchRemoteEgressApplyLocal
+	}
```

Comment:
I am not a fan of this at all. At this point on wards, you could just have different methods.